phac-nml/irida-next
# Update: Membership source

## What does this PR do and why?
_Describe in detail what your merge request does and why._

Updated members tables for groups and projects to include a `source` column which displays if the member is a `Direct Member` or displays the group name with a link to it, if the membership is inherited from a parent group.

Resolves #111

## Screenshots or screen recordings
_Screenshots are required for UI changes, and strongly recommended for all other pull requests._

![image](https://github.com/phac-nml/irida-next/assets/25466211/41340b77-107a-49dc-b616-379ff7407d58)

![image](https://github.com/phac-nml/irida-next/assets/25466211/bad191fc-59da-497d-900d-808af627a72d)

![image](https://github.com/phac-nml/irida-next/assets/25466211/9608a036-d15f-4328-b089-42a7290f1fb4)

## How to set up and validate locally
_Numbered steps to set up and validate the change are strongly suggested._

1. Login as `admin@email.com`
2. Go the the `Bacillus` group and view it's members
3. You should see a listing of the members and in the source column you should see `Direct Member`
4. Now go to the `Bacillus cerus` group and view it's members
5. You should see the same members listed from step 3 but now under the source column there should be a link to the `Bacillus` group in the source column for each member
6. Go to the `Outbreak 2021` project under the `Bacillus cerus` group
7. You should see user6@email.com, user7@email.com, and user8@email.com with `Direct Member` listed under the source column. These members are also a part of the group but they are also added as direct members to the project. `admin@email.com` and `user9@email.com` have the group link from which they are inherited. You will need to add `user9@email.com` to the `Bacillus` group to see them in the listing for the project.
8. Try this out with different scenarios such as where the user is a member of a parent group, and is a direct member of a subgroup within the parent group, and is a direct member of a project within the subgroup. Try removing the member as a direct member from the project. Under the project members table you should see the user is still in the listing but inherited from the subgroup with a link to it. Now try removing them from the subgroup. View the members for the subgroup and project. You should see the member is inherited from the parent group with a link to it.

**Note: An inherited member's access level can only be updated via the group that they are inherited from. Only a direct member's access level can be updated for both groups and projects**

## PR acceptance checklist
This checklist encourages us to confirm any changes have been analyzed to reduce risks in quality, performance, reliability, security, and maintainability.
